src/linear_algebra/quadratic_form.lean

```diff
@@ -0,0 +1,290 @@
+/-
+Copyright (c) 2020 Anne Baanen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Anne Baanen
+
+Quadratic forms over modules.
+-/
+
+import linear_algebra.bilinear_form
+import linear_algebra.determinant
+import linear_algebra.special_linear_group
+
+/-!
+# Quadratic forms
+
+This file defines quadratic forms over a `R`-module `M`.
+A quadratic form is a map `Q : M → R` such that
+  (`map_smul`) `Q (a • x) = a * a * Q x`
+  (`polar_...`) The map `polar Q := λ x y, Q (x + y) - Q x - Q y` is bilinear.
+They come with a scalar multiplication, `(a • Q) x = Q (a • x) = a * a * Q x`,
+and composition with linear maps `f`, `Q.comp f x = Q (f x)`.
+
+## Main definitions
+
+ * `quadratic_form.associated`: associated bilinear form
+ * `quadratic_form.pos_def`: positive definite quadratic forms
+ * `quadratic_form.discr`: discriminant of a quadratic form
+
+## Main statements
+
+ * `quadratic_form.associated_left_inverse`,
+ * `quadratic_form.associated_right_inverse`: in a commutative ring where 2 has
+  an inverse, there is a correspondence between quadratic forms and symmetric
+  bilinear forms
+
+## References
+
+ * https://en.wikipedia.org/wiki/Quadratic_form
+ * https://en.wikipedia.org/wiki/Discriminant#Quadratic_forms
+
+## Tags
+
+quadratic form, homogeneous polynomial, quadratic polynomial
+-/
+
+universes u v w
+variables {R : Type u} {M : Type v} [add_comm_group M] [ring R]
+variables {R₁ : Type u} [comm_ring R₁] [module R₁ M]
+
+/-- Up to a factor 2, `Q.polar` is the associated bilinear form for a quadratic form `Q`. -/
+def quadratic_form.polar (f : M → R) (x y : M) :=
+f (x + y) - f x - f y
+
+variables [module R M]
+
+open quadratic_form
+/-- A quadratic form over a module. -/
+structure quadratic_form (R : Type u) (M : Type v) [ring R] [add_comm_group M] [module R M] :=
+(map : M → R)
```

Comment:
In the rest of the library, the carrier function of a bundled map is usually called `to_fun`.
```suggestion
(to_fun : M → R)
```